### DISCUSSION TITLE
creating scene, adding input, sizing in scene

### DISCUSSION THREAD

**Discussion by davehorner** [dnaka91/obws#35] [General]
Hello,

Thanks for the information in: https://github.com/dnaka91/obws/issues/34
Figured I'd move it to a discussion as it isn't a bug or PR,

I've got code that I'm using;
https://github.com/dnaka91/obws/commit/629530d302274b417d5bcd6b34ebb1a2ea9784eb
its not ready for sharing but it is helpful for communicating.

The first time, it creates the scene/source/input correctly.  The subsequent run, it does list the settings for the source and then it exits Error: API error: ResourceAlreadyExists.   I'm calling client.inputs().remove(source).await.ok(); and it seems to work, do I need to do something else after the remove or why might I be getting this ResourceAlreadyExists.

The code is working.  Wondering how I size the input in scene on screen.
client.inputs().settings::<serde_json::Value>(source).await.ok();
was very helpful.  wondering what I need to be looking at to watch and set layout.

anyway; thanks again.

**Comment by davehorner** [comment]
I found set_transform and it scales fine.
```
    let mut t: SceneItemTransform = Default::default();
    t.scale=Some(Scale{x:Some(2.0),y:Some(2.0)});
client.scene_items().set_transform(
     SetTransform {item_id: item_id,scene: &scene_name, transform:t});
```
In the UI you can center to screen, fit to screen, etc and those actions translate into a computed transform.  Is there a way to perform these actions and not have to compute the transform via construction/re-implement these transforms? 

I played a bit more https://github.com/dnaka91/obws/commit/5e6c2dbce4c312c2902ab421b1f880436fd616cc trying to change the source input if its already there... set private settings didn't seem right, was looking at client.inputs().set_settings but that also didn't seem to work for me. cheers.

**Reply by dnaka91** [reply]
Sadly, obs-websocket doesn't provide any API for these convenience functions. Therefore, you'd have to build them yourself from scratch. Basically means getting the screen size, then the item's size and calculate the offset.

I haven't started anything to support their batching API yet, which would allow you to bundle that all into a single batch of requests without all the forth and back between the client and server. Pretty interesting actually, because they have a concept of referencing the output of the previous command as an input for the following command and other neat tricks.

Just haven't found a good way to build a type safe system around that, which is flexible but as typed out as possible at the same time. Maybe I could build a raw API as a basic building block, but that wouldn't provide any safety measures.

**Reply by dnaka91** [reply]
> wondering what I need to be looking at to watch and set layout.

Looks like you already found the right spot to set the layout with `set_transform`. To watch for any changes you can react on the [ScenItemTransformChanged](https://docs.rs/obws/0.11.5/obws/events/enum.Event.html#variant.SceneItemTransformChanged) event, but have to be careful to not get stuck in a loop from your own transform changes.

What I mean is, IIRC the event fires even for your own commands, so you need to make sure to skip any logic based on that event, if it was triggered from your own `set_transform` calls.
